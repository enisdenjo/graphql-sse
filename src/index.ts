export * from './common';
export * from './handler';
export * from './client';
