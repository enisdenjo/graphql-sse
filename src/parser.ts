/**
 *
 * parser
 *
 */

import {
  StreamMessage,
  validateStreamEvent,
  parseStreamData,
  StreamEvent,
} from './common';

enum ControlChars {
  NewLine = 10,
  CarriageReturn = 13,
  Space = 32,
  Colon = 58,
}

/**
 * HTTP response chunk parser for graphql-sse's event stream messages.
 *
 * Reference: https://github.com/Azure/fetch-event-source/blob/main/src/parse.ts
 *
 * @private
 */
export function createParser<ForID extends boolean>(): (
  chunk: Uint8Array,
) => StreamMessage<ForID, StreamEvent>[] | void {
  let buffer: Uint8Array | undefined;
  let position: number; // current read position
  let fieldLength: number; // length of the `field` portion of the line
  let discardTrailingNewline = false;
  let message = { event: '', data: '' };
  let pending: StreamMessage<ForID, StreamEvent>[] = [];
  const decoder = new TextDecoder();

  return function parse(chunk) {
    if (buffer === undefined) {
      buffer = chunk;
      position = 0;
      fieldLength = -1;
    } else {
      const concat = new Uint8Array(buffer.length + chunk.length);
      concat.set(buffer);
      concat.set(chunk, buffer.length);
      buffer = concat;
    }

    const bufLength = buffer.length;
    let lineStart = 0; // index where the current line starts
    while (position < bufLength) {
      if (discardTrailingNewline) {
        if (buffer[position] === ControlChars.NewLine) {
          lineStart = ++position; // skip to next char
        }
        discardTrailingNewline = false;
      }

      // look forward until the end of line
      let lineEnd = -1; // index of the \r or \n char
      for (; position < bufLength && lineEnd === -1; ++position) {
        switch (buffer[position]) {
          case ControlChars.Colon:
            if (fieldLength === -1) {
              // first colon in line
              fieldLength = position - lineStart;
            }
            break;
          // \r case below should fallthrough to \n:
          case ControlChars.CarriageReturn:
            discardTrailingNewline = true;
          // eslint-disable-next-line no-fallthrough
          case ControlChars.NewLine:
            lineEnd = position;
            break;
        }
      }

      if (lineEnd === -1) {
        // end of the buffer but the line hasn't ended
        break;
      } else if (lineStart === lineEnd) {
        // empty line denotes end of incoming message
        if (message.event || message.data) {
          // NOT a server ping (":\n\n")
          if (!message.event) throw new Error('Missing message event');
          const event = validateStreamEvent(message.event);
          const data = parseStreamData<ForID, StreamEvent>(event, message.data);
          pending.push({
            event,
            data,
          });
          message = { event: '', data: '' };
        }
      } else if (fieldLength > 0) {
        // end of line indicates message
        const line = buffer.subarray(lineStart, lineEnd);

        // exclude comments and lines with no values
        // line is of format "<field>:<value>" or "<field>: <value>"
        // https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation
        const field = decoder.decode(line.subarray(0, fieldLength));
        const valueOffset =
          fieldLength + (line[fieldLength + 1] === ControlChars.Space ? 2 : 1);
        const value = decoder.decode(line.subarray(valueOffset));

        switch (field) {
          case 'event':
            message.event = value;
            break;
          case 'data':
            // append the new value if the message has data
            message.data = message.data ? message.data + '\n' + value : value;
            break;
        }
      }

      // next line
      lineStart = position;
      fieldLength = -1;
    }

    if (lineStart === bufLength) {
      // finished reading
      buffer = undefined;
      const messages = [...pending];
      pending = [];
      return messages;
    } else if (lineStart !== 0) {
      // create a new view into buffer beginning at lineStart so we don't
      // need to copy over the previous lines when we get the new chunk
      buffer = buffer.subarray(lineStart);
      position -= lineStart;
    }
  };
}
