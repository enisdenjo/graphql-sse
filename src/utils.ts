/**
 *
 * utils
 *
 */

/** @private */
export function isObject(val: unknown): val is Record<PropertyKey, unknown> {
  return typeof val === 'object' && val !== null;
}
