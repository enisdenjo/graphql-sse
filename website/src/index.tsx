import React from 'react';
import clsx from 'clsx';
import { FiGithub, FiPlay } from 'react-icons/fi';
import {
  TbArrowMergeBoth,
  TbArrowUpCircle,
  TbChefHat,
  TbPlugConnected,
  TbRepeat,
  TbServer,
} from 'react-icons/tb';
import { BsBoxSeam } from 'react-icons/bs';
import { FaCreativeCommonsZero } from 'react-icons/fa';
import { Anchor, Image } from '@theguild/components';
import Link from 'next/link';

const classes = {
  button: {
    gray: 'inline-block bg-gray-200 hover:bg-gray-300 dark:bg-gray-800 dark:text-gray-300 dark:hover:bg-gray-700 text-gray-600 px-6 py-3 rounded-lg font-medium shadow-sm',
    emerald:
      'inline-block bg-emerald-200 hover:bg-emerald-300 dark:bg-emerald-800 dark:text-emerald-300 dark:hover:bg-emerald-700 text-emerald-600 px-6 py-3 rounded-lg font-medium shadow-sm',
  },
  link: {
    blue: 'text-blue-600 hover:text-blue-800 dark:hover:text-blue-400',
    emerald:
      'text-emerald-600 hover:text-emerald-800 dark:hover:text-emerald-400',
  },
};

const gradients: [string, string][] = [
  ['#06b6d4', '#0e7490'], // cyan
  ['#f97316', '#c2410c'], // orange
  ['#10b981', '#047857'], // emerald
];

function pickGradient(i: number) {
  const gradient = gradients[i % gradients.length];
  if (!gradient) {
    throw new Error('No gradient found');
  }
  return gradient;
}

export function Index() {
  return (
    <>
      <Wrapper>
        <div className="container py-20 sm:py-24 lg:py-32">
          <h1 className="max-w-screen-md mx-auto font-extrabold text-5xl sm:text-5xl lg:text-6xl text-center bg-gradient-to-r from-green-700 to-emerald-400 dark:from-green-700 dark:to-emerald-400 bg-clip-text text-transparent !leading-tight">
            GraphQL SSE
          </h1>
          <p className="max-w-screen-sm mx-auto mt-6 text-2xl text-gray-600 text-center dark:text-gray-400">
            Zero-dependency, HTTP/1 safe, simple, spec compliant server and
            client
          </p>
          <div className="mt-10 flex flex-col sm:flex-row items-center justify-center gap-4">
            <Anchor
              className={clsx(
                classes.button.gray,
                'flex flex-row gap-2 items-center',
              )}
              href="/get-started"
            >
              <FiPlay /> Get Started
            </Anchor>
            <Anchor
              className={clsx(
                classes.button.gray,
                'flex flex-row gap-2 items-center',
              )}
              href="/recipes"
            >
              <TbChefHat /> Recipes
            </Anchor>
            <Anchor
              className={clsx(
                classes.button.gray,
                'flex flex-row gap-2 items-center',
              )}
              href="https://github.com/enisdenjo/graphql-sse"
            >
              <FiGithub /> GitHub
            </Anchor>
          </div>
        </div>
      </Wrapper>

      <Wrapper>
        <Feature
          gradient={0}
          title="Spec Compliant"
          description={
            <div className="flex flex-col gap-y-12">
              <p className="text-center mb-6">
                As a reference implementation, the library is fully compliant
                with the{' '}
                <Link
                  href="https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md"
                  className={classes.link.blue}
                >
                  GraphQL over SSE (Server-Sent Events) spec
                </Link>
              </p>
              <div className="flex flex-col gap-y-12">
                <FeatureHighlights
                  textColor={gradients[0][0]}
                  highlights={[
                    {
                      icon: <TbArrowMergeBoth size={36} />,
                      title: 'Single connection mode',
                      description:
                        'Safe for HTTP/1 servers and subscription heavy apps',
                    },
                    {
                      icon: <TbPlugConnected size={36} />,
                      title: 'Distinct connection mode',
                      description:
                        'Each connection is a subscription on its own',
                    },
                    {
                      icon: <TbArrowUpCircle size={36} />,
                      title: 'Upgrade over regular SSE',
                      description:
                        'Contains improvements for clarity and stability',
                    },
                  ]}
                />
              </div>
            </div>
          }
        />
      </Wrapper>

      <Wrapper>
        <Feature
          gradient={1}
          title="Server and Client"
          description={
            <div className="flex flex-col gap-y-12">
              <p className="text-center mb-6">
                Single library, but both the server and the client is included
              </p>
              <div className="flex flex-col gap-y-12">
                <FeatureHighlights
                  textColor={gradients[1][0]}
                  highlights={[
                    {
                      icon: <FaCreativeCommonsZero size={36} />,
                      title: 'Zero-dependency',
                      description: 'Library contains everything it needs',
                    },
                    {
                      link: 'https://bundlephobia.com/package/graphql-sse',
                      icon: <BsBoxSeam size={36} />,
                      title: 'Still small bundle',
                      description:
                        'Thanks to tree-shaking and module separation',
                    },
                    {
                      icon: <TbRepeat size={36} />,
                      title: 'Smart retry strategies',
                      description:
                        'Customize the reconnection process to your desires',
                    },
                    {
                      icon: <TbServer size={36} />,
                      title: 'Run everywhere',
                      description:
                        'Completely server agnostic, run them both anywhere',
                    },
                  ]}
                />
              </div>
            </div>
          }
        />
      </Wrapper>

      <Wrapper>
        <Feature
          gradient={2}
          icon={<TbChefHat />}
          title="Recipes"
          description={
            <div className="flex flex-col gap-y-12">
              <p className="text-center mb-6">
                Short and concise code snippets for starting with common
                use-cases
              </p>
              <div className="flex flex-col gap-y-12">
                <ul className="my-0">
                  <li>
                    <Link
                      className={classes.link.emerald}
                      href="/recipes#client-usage-with-async-iterator"
                    >
                      Client usage with AsyncIterator
                    </Link>
                  </li>
                  <li>
                    <Link
                      className={classes.link.emerald}
                      href="/recipes#client-usage-with-relay"
                    >
                      Client usage with Relay
                    </Link>
                  </li>
                  <li>
                    <Link
                      className={classes.link.emerald}
                      href="/recipes#server-handler-usage-with-custom-context-value"
                    >
                      Server handler usage with custom context value
                    </Link>
                  </li>
                  <li>
                    <Link
                      className={classes.link.emerald}
                      href="/recipes#server-handler-and-client-usage-with-persisted-queries"
                    >
                      Server handler and client usage with persisted queries
                    </Link>
                  </li>
                  <li className="flex mt-2">
                    <Anchor
                      className={clsx(
                        classes.button.emerald,
                        'flex flex-row gap-2 items-center',
                      )}
                      href="/recipes"
                    >
                      And many more...
                    </Anchor>
                  </li>
                </ul>
              </div>
            </div>
          }
        />
      </Wrapper>
    </>
  );
}

function Wrapper({ children }: { children: React.ReactNode }) {
  return (
    <div
      className={clsx(
        'w-full',
        'py-24',
        'odd:bg-gray-50',
        'odd:dark:bg-gray-900',
        'even:bg-white',
        'even:dark:bg-black',
      )}
    >
      {children}
    </div>
  );
}

function Feature({
  icon,
  title,
  description,
  children,
  image,
  gradient,
  flipped,
}: {
  children?: React.ReactNode;
  icon?: React.ReactNode;
  title: string;
  description: React.ReactNode;
  highlights?: {
    title: string;
    description: React.ReactNode;
    icon?: React.ReactNode;
  }[];
  image?: string;
  gradient: number;
  flipped?: boolean;
}) {
  const [start, end] = pickGradient(gradient);

  return (
    <div className="container box-border px-6 mx-auto flex flex-col gap-y-24">
      <div
        className={clsx(
          'flex flex-col gap-24 md:gap-12 lg:gap-24 justify-center items-stretch',
          flipped ? 'md:flex-row-reverse' : 'md:flex-row',
        )}
      >
        <div
          className={clsx(
            'flex flex-col gap-4 w-full md:w-3/5 lg:w-2/5 flex-shrink-0',
            !image && 'items-center',
          )}
        >
          {icon && (
            <div className="text-5xl" style={{ color: start }}>
              {icon}
            </div>
          )}
          <h2
            className={clsx(
              'font-semibold text-5xl bg-clip-text text-transparent dark:text-transparent leading-normal',
              !image && 'text-center',
            )}
            style={{
              backgroundImage: `linear-gradient(-70deg, ${end}, ${start})`,
            }}
          >
            {title}
          </h2>
          <div className="text-lg text-gray-600 dark:text-gray-400 leading-7">
            {description}
          </div>
        </div>
        {image && (
          <div
            className="rounded-3xl overflow-hidden p-8 flex-grow flex flex-col justify-center items-stretch relative"
            style={{
              backgroundImage: `linear-gradient(70deg, ${start}, ${end})`,
            }}
          >
            <Image
              src={image}
              fill
              className="rounded-xl mx-auto"
              placeholder="empty"
              alt={title}
            />
          </div>
        )}
      </div>
      {children}
    </div>
  );
}

function FeatureHighlights({
  highlights,
  textColor,
}: {
  textColor?: string;
  highlights?: {
    title: string;
    description: React.ReactNode;
    icon?: React.ReactNode;
    link?: string;
  }[];
}) {
  if (!Array.isArray(highlights)) {
    return null;
  }

  return (
    <>
      {highlights.map(({ title, description, icon, link }) => {
        const Comp = link ? Anchor : 'div';
        return (
          <Comp
            key={title}
            className="flex flex-row md:flex-col lg:flex-row flex-1 gap-4"
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            {...(link && ({ href: link } as any))}
          >
            {icon && (
              <div
                className="flex-shrink-0"
                style={textColor ? { color: textColor } : {}}
              >
                {icon}
              </div>
            )}
            <div className="text-black dark:text-white">
              <h3
                className={clsx('text-xl font-semibold', !icon && 'text-lg')}
                style={textColor ? { color: textColor } : {}}
              >
                {title}
              </h3>
              <p
                className={clsx(
                  'text-gray-600 dark:text-gray-400',
                  !icon && 'text-sm',
                )}
              >
                {description}
              </p>
            </div>
          </Comp>
        );
      })}
    </>
  );
}
