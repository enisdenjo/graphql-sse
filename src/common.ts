/**
 *
 * common
 *
 */

import type { GraphQLError } from 'graphql';
import { isObject } from './utils';

/**
 * Header key through which the event stream token is transmitted
 * when using the client in "single connection mode".
 *
 * Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode
 *
 * @category Common
 */
export const TOKEN_HEADER_KEY = 'x-graphql-event-stream-token' as const;

/**
 * URL query parameter key through which the event stream token is transmitted
 * when using the client in "single connection mode".
 *
 * Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode
 *
 * @category Common
 */
export const TOKEN_QUERY_KEY = 'token' as const;

/**
 * Parameters for GraphQL's request for execution.
 *
 * Reference: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#request
 *
 * @category Common
 */
export interface RequestParams {
  operationName?: string;
  query: string;
  variables?: Record<string, unknown>;
  extensions?: Record<string, unknown>;
}

/**
 * Represents a message in an event stream.
 *
 * Read more: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#Event_stream_format
 *
 * @category Common
 */
export interface StreamMessage<ForID extends boolean, E extends StreamEvent> {
  // id?: string; might be used in future releases for connection recovery
  event: E;
  data: ForID extends true ? StreamDataForID<E> : StreamData<E>;
  // retry?: number; ignored since graphql-sse implements custom retry strategies
}

/** @category Common */
export type StreamEvent = 'next' | 'complete';

/** @category Common */
export function validateStreamEvent(e: unknown): StreamEvent {
  e = e as StreamEvent;
  if (e !== 'next' && e !== 'complete')
    throw new Error(`Invalid stream event "${e}"`);
  return e;
}

/** @category Common */
export function print<ForID extends boolean, E extends StreamEvent>(
  msg: StreamMessage<ForID, E>,
): string {
  let str = `event: ${msg.event}\ndata:`;
  if (msg.data) {
    str += ' ';
    str += JSON.stringify(msg.data);
  }
  str += '\n\n';
  return str;
}

/** @category Common */
export interface ExecutionResult<
  Data = Record<string, unknown>,
  Extensions = Record<string, unknown>,
> {
  errors?: ReadonlyArray<GraphQLError>;
  data?: Data | null;
  hasNext?: boolean;
  extensions?: Extensions;
}

/** @category Common */
export interface ExecutionPatchResult<
  Data = unknown,
  Extensions = Record<string, unknown>,
> {
  errors?: ReadonlyArray<GraphQLError>;
  data?: Data | null;
  path?: ReadonlyArray<string | number>;
  label?: string;
  hasNext: boolean;
  extensions?: Extensions;
}

/** @category Common */
export type StreamData<E extends StreamEvent> = E extends 'next'
  ? ExecutionResult | ExecutionPatchResult
  : E extends 'complete'
  ? null
  : never;

/** @category Common */
export type StreamDataForID<E extends StreamEvent> = E extends 'next'
  ? { id: string; payload: ExecutionResult | ExecutionPatchResult }
  : E extends 'complete'
  ? { id: string }
  : never;

/** @category Common */
export function parseStreamData<ForID extends boolean, E extends StreamEvent>(
  e: E,
  data: string,
) {
  if (data) {
    try {
      data = JSON.parse(data);
    } catch {
      throw new Error('Invalid stream data');
    }
  }

  if (e === 'next' && !data)
    throw new Error('Stream data must be an object for "next" events');

  return (data || null) as ForID extends true
    ? StreamDataForID<E>
    : StreamData<E>;
}

/**
 * A representation of any set of values over any amount of time.
 *
 * @category Common
 */
export interface Sink<T = unknown> {
  /** Next value arriving. */
  next(value: T): void;
  /** An error that has occured. This function "closes" the sink. */
  error(error: unknown): void;
  /** The sink has completed. This function "closes" the sink. */
  complete(): void;
}

/**
 * Checkes whether the provided value is an async iterable.
 *
 * @category Common
 */
export function isAsyncIterable<T>(val: unknown): val is AsyncIterable<T> {
  return typeof Object(val)[Symbol.asyncIterator] === 'function';
}

/**
 * Checkes whether the provided value is an async generator.
 *
 * @category Common
 */
export function isAsyncGenerator<T>(val: unknown): val is AsyncGenerator<T> {
  return (
    isObject(val) &&
    typeof Object(val)[Symbol.asyncIterator] === 'function' &&
    typeof val.return === 'function' &&
    typeof val.throw === 'function' &&
    typeof val.next === 'function'
  );
}
