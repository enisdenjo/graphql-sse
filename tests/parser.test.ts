import { it } from 'vitest';
import { createParser } from '../src/parser';

const encoder = new TextEncoder();

it('should parse whole message', ({ expect }) => {
  const parse = createParser();

  // with space
  expect(
    parse(encoder.encode('event: next\ndata: { "iAm": "data" }\n\n')),
  ).toMatchSnapshot();

  // no space
  expect(
    parse(encoder.encode('event:next\ndata:{ "iAm": "data" }\n\n')),
  ).toMatchSnapshot();

  // no data
  expect(parse(encoder.encode('event: complete\n\n'))).toMatchSnapshot();
});

it('should parse message with prepended ping', ({ expect }) => {
  const parse = createParser();

  expect(
    parse(encoder.encode(':\n\nevent:next\ndata:{ "iAm": "data" }\n\n')),
  ).toMatchSnapshot();
});

it('should parse chunked message', ({ expect }) => {
  const parse = createParser();

  parse(encoder.encode('even'));
  parse(encoder.encode(''));
  parse(encoder.encode('t: ne'));
  parse(encoder.encode('xt\nda'));
  parse(encoder.encode('ta:{'));
  parse(encoder.encode(''));
  parse(encoder.encode(' "iAm'));
  const msg = parse(encoder.encode('": "data" }\n\n'));

  expect(msg).toMatchSnapshot();
});

it('should parse message whose lines are separated by \\r\\n', ({ expect }) => {
  const parse = createParser();

  const msg = parse(
    encoder.encode('event: next\r\ndata: { "iAm": "data" }\r\n\r\n'),
  );

  expect(msg).toMatchSnapshot();
});

it('should ignore comments', ({ expect }) => {
  const parse = createParser();

  const msg = parse(
    encoder.encode(': I am comment\nevent: next\ndata: { "iAm": "data" }\n\n'),
  );

  expect(msg).toMatchSnapshot();
});

it('should accept valid events only', ({ expect }) => {
  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: done\ndata: {}\n\n'));
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: value\ndata: {}\n\n'));
  }).toThrowErrorMatchingSnapshot();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: next\ndata: {}\n\n'));
  }).not.toThrow();

  expect(() => {
    const parse = createParser();
    parse(encoder.encode('event: complete\ndata: {}\n\n'));
  }).not.toThrow();
});

it('should ignore server pings', ({ expect }) => {
  const parse = createParser();

  expect(parse(encoder.encode(':\n\n'))).toEqual([]);
});

it('should parse multiple messages from one chunk', ({ expect }) => {
  const parse = createParser();

  expect(
    parse(
      encoder.encode(
        'event: next\ndata: {}\n\n' + 'event: next\ndata: { "no": "data" }\n\n',
      ),
    ),
  ).toMatchSnapshot();

  expect(
    parse(
      encoder.encode(
        'event: next\ndata: { "almost": "done" }\n\n' +
          'event: complete\ndata: {}\n\n',
      ),
    ),
  ).toMatchSnapshot();
});
