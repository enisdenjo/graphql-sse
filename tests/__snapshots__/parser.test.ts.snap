// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should accept valid events only 1`] = `"Invalid stream event \\"done\\""`;

exports[`should accept valid events only 2`] = `"Invalid stream event \\"value\\""`;

exports[`should ignore comments 1`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse chunked message 1`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse message whose lines are separated by \\r\\n 1`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse message with prepended ping 1`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse multiple messages from one chunk 1`] = `
[
  {
    "data": {},
    "event": "next",
  },
  {
    "data": {
      "no": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse multiple messages from one chunk 2`] = `
[
  {
    "data": {
      "almost": "done",
    },
    "event": "next",
  },
  {
    "data": {},
    "event": "complete",
  },
]
`;

exports[`should parse whole message 1`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse whole message 2`] = `
[
  {
    "data": {
      "iAm": "data",
    },
    "event": "next",
  },
]
`;

exports[`should parse whole message 3`] = `
[
  {
    "data": null,
    "event": "complete",
  },
]
`;
